### PR TITLE
BUG: groupby sum, mean, var should always be floats

### LINKED ISSUE #41137: BUG: groupby mean/median/var should always be float values
This is value-dependent behavior:

```
print(DataFrame({'a': [1, 3]}).groupby([1, 1]).mean())
```

results in the integer 2, whereas

```
print(DataFrame({'a': [1, 3]}).mean())
```

results in 2.0.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #41137
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Built on top of #40790

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -290,10 +290,13 @@ def get_result_dtype(self, dtype: DtypeObj) -> DtypeObj:
                 return np.dtype(np.int64)
             elif isinstance(dtype, (BooleanDtype, _IntegerDtype)):
                 return Int64Dtype()
-        elif how in ["mean", "median", "var"] and isinstance(
-            dtype, (BooleanDtype, _IntegerDtype)
-        ):
-            return Float64Dtype()
+        elif how in ["mean", "median", "var"]:
```
Comment (jreback) [inline pandas/core/groupby/ops.py:293]: this entire routine should be in pandas/core/dtypes/common.py or similar. I think we had an issue about this.

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 2]
```diff
@@ -442,6 +441,57 @@ def test_bool_agg_dtype(op):
...
+@pytest.mark.parametrize(
+    "input_dtype", ["bool", "int32", "int64", "float32", "float64"]
+)
+@pytest.mark.parametrize(
+    "result_dtype", ["bool", "int32", "int64", "float32", "float64"]
+)
+@pytest.mark.parametrize("method", ["apply", "aggregate", "transform"])
+def test_callable_result_dtype_frame(
+    keys, agg_index, input_dtype, result_dtype, method
+):
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:460]: do these cover nullable types?

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. +1 generally on this behavior.

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/groupby/ops.py:293]: This was just moved from dtyes/cast.py in #41065

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/tests/groupby/aggregate/test_aggregate.py:460]: This is still not properly handled for EAs - issue is #37494

**Comment by jreback** [comment]
looks good. can you add a whatsnew note. might consider a sub-section as this is a 'bigger' bug fix but either way.

**Comment by rhshadrach** [comment]
@jreback Thanks - whatsnew section added.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rhshadrach

File: pandas/tests/resample/test_datetime_index.py [hunk 3]
```diff
@@ -1687,7 +1687,7 @@ def f(data, add_arg):
 
     # Testing dataframe
     df = DataFrame({"A": 1, "B": 2}, index=date_range("2017", periods=10))
-    result = df.groupby("A").resample("D").agg(f, multiplier)
+    result = df.groupby("A").resample("D").agg(f, multiplier).astype(float)
     expected = df.groupby("A").resample("D").mean().multiply(multiplier)
     # TODO: GH 41137
```
Comment (jbrockmendel) [inline pandas/tests/resample/test_datetime_index.py:1692]: @rhshadrach does this PR take care of the TODO comment here?

Comment (rhshadrach) [reply pandas/tests/resample/test_datetime_index.py:1692]: Yes - thanks, I've opened #44374

**Pull request merged**